docs: added content for securing operations (#2172)
asyncapi/website: pages/docs/concepts/asyncapi-document/securing-operations.md

---
title: 'Securing Operations'
weight: 120
---

In an AsyncAPI document, the concept of server security implies that the security measures defined at the server level apply to all operations within all channels by default. 

## Features of Security in AsyncAPI Documents

- The security requirements specified at the server level are enforced consistently across the entire asyncapi document.

> There may be situations where certain operations within specific channels require different security measures than the default server-level settings. To accommodate for such scenarios, AsyncAPI document allows you to use the `security` property at the `operation` level.

Reviewer: another situation is when someone just do not use `server` feature from AsyncAPI, they have just channels and operations. So they need a way to specify security of the operation

example https://github.com/asyncapi/spec/blob/next-major-spec/examples/operation-security.yml

Author: defined both ways of adding security to operations

![image](https://github.com/asyncapi/website/assets/78029145/d0f19cde-4840-4907-ae2c-7e1370343449)

> By using the `security` property at the `operation` level, you can override the default server-level security and define unique security requirements for individual operations. Thereby also giving you the flexibility to tailor the security measures to the specific needs of each operation, even if they differ from the broader server-level settings.

Reviewer: it is not much about overriding, please have a look at text from the spec
![Screenshot 2023-10-12 at 12 55 27](https://github.com/asyncapi/website/assets/6995927/b41f5cc1-6b4c-43ce-8b69-ad01b22040c4)

Author: - overriding sentence removed
- added objects under `security` in v3

![image](https://github.com/asyncapi/website/assets/78029145/984f7549-ca8d-40c2-bfcd-e0106ab4bdfd)

## Specifying Security Requirement 

To specify different security requirements for a specific operation, you can include the `security` property within the operation's definition. 
The security property is an array where you can define one or more security requirement objects.

> For example, let's say you have an AsyncAPI document with a channel called users and two operations within that channel: createUser and getUser. 

Reviewer: please use markdown code formatting with back tick when writing some values names, like createUser should be `createUser`

Reviewer: btw, operations are not within the channel in v3

Author: - markdown format used and corrected
- changed the example from v2 to v3 

![image](https://github.com/asyncapi/website/assets/78029145/7d425a81-62aa-4e88-8996-858190e58cb7)

> The server-level security is set to use API key authentication for all operations within all channels. 

Reviewer: you cannot enforce as it is not about overriding, 

it is like providing alternative for what is already in the server. Of course you can do a server without security, then have different operations and channels on the same server, each operation with different security - still, it is not overriding/enforcing if you know what I mean

Author: I am unable to understand how to improve the content as per your suggestions.
could you please attach some examples or maybe give me a bit more info about the same?

Author: > you cannot enforce as it is not about overriding,
> 
> it is like providing alternative for what is already in the server. Of course you can do a server without security, then have different operations and channels on the same server, each operation with different security - still, it is not overriding/enforcing if you know what I mean

I have removed the line that stated about enforcing/overriding, though not sure if I still was able to resolved the suggestion.

However, you want to enforce OAuth2 authentication specifically for the getUser operation.

You can achieve this by including the security property at the operation level, as shown in the following example:

> ```yaml

Reviewer: this is v2 example

Author: replaced with v3 example

![image](https://github.com/asyncapi/website/assets/78029145/ffb685c2-5151-434f-a239-a10daebf7757)

> In an AsyncAPI document, the concept of server security implies that the security measures defined at the server level apply to all operations within all channels by default. 

Reviewer 2: Remember to add the title of the document and page weight for every document you add😁. Adding the title and weight is important as the page weight is used to determine the order of the documents. Make sense?
```suggestion
---
title: 'Securing Operations'
weight: 120
---

In an AsyncAPI document, the concept of server security implies that the security measures defined at the server level apply to all operations within all channels by default.

> ## Features of Security in AsyncAPI Documents

Reviewer 2: The document goes under the sub-section or sub-content bucket called "AsyncAPI Document", so by default all the documents in the section talk about how a certain concept works in AsyncAPI Document. So we can shorten the sub section here :) 

So, can you rewrite the heading of the section while keeping it concise yet conveying the message?😄

Something among the lines of "Features of Security" or "Security Features" etc.

Author: shortened the part according to your suggestions

![image](https://github.com/asyncapi/website/assets/78029145/5faed145-f170-4bef-beed-fc7db11af6dc)

> - The security requirements specified at the server level are enforced consistently across the entire API.

Reviewer 2: AsyncAPI document does not do anything until instantiated, so across the entire specification or asyncapi document seems more suitable and less confusing for a reader, don't you think?
```suggestion
- The security requirements specified at the server level are enforced consistently across the entire asyncapi document.
```

- There may be situations where certain operations within specific channels require different security measures than the default server-level settings.

- To accommodate such scenarios, AsyncAPI allows you to use the security property at the operation level.

- By using the security property at the operation level, you can override the default server-level security and define unique security requirements for individual operations.

- Gives the flexibility to tailor the security measures to the specific needs of each operation, even if they differ from the broader server-level settings.

- When you include the security property at the operation level, it takes precedence over the server-level security settings for that particular operation.

> - The security measures specified at the operation level will be applied instead of the default server-level security.

Reviewer 2: Looks like the first 3 points here are actually addressing 1 whole point, so it is more easy to read and follow if they are written in a para-flow. Rest of the points also seem to be extension points. So it makes more sense to phrase the points in paragraphs making the flow and understanding easier.

The last point seems repetitive or redundant as it conveys the same point as the points above it, don't you think?
```suggestion
There may be situations where certain operations within specific channels require different security measures than the default server-level settings. To accommodate for such scenarios, AsyncAPI document allows you to use the `security` property at the `operation` level.

By using the `security` property at the `operation` level, you can override the default server-level security and define unique security requirements for individual operations. Thereby also giving you the flexibility to tailor the security measures to the specific needs of each operation, even if they differ from the broader server-level settings.
```

## Specifying Security Requirement 

> To specify different security requirements for a specific operation, you can include the security property within the operation's definition. 

Reviewer 2: Use backticks throughout the document where required :)
```suggestion
To specify different security requirements for a specific operation, you can include the `security` property within the operation's definition. 
```

The security property is an array where you can define one or more security requirement objects.

For example, let's say you have an AsyncAPI document with a channel called users and two operations within that channel: createUser and getUser. 
The server-level security is set to use API key authentication for all operations within all channels. 
However, you want to enforce OAuth2 authentication specifically for the getUser operation.

You can achieve this by including the security property at the operation level, as shown in the following example:

> ```

Reviewer 2: ```suggestion
```yaml
```

channels:
  users:
    publish:
      summary: Creates a user
      operationId: createUser
      message:
        ...
    subscribe:
      summary: Retrieves user information
      operationId: getUser
      message:
        ...
      security:
        - OAuth2: []
```

> In this example, the security property is added under the getUser operation, indicating that the OAuth2 security requirement should be applied to that specific operation within the user's channel. The empty array [] signifies that no additional configuration is needed for the OAuth2 security mechanism.

Reviewer 2: "this" is ambiguous and antecedent. 
```suggestion
In the above example, the `security` property is added under the `getUser` operation, indicating that the OAuth2 security requirement should be applied to that specific operation within the user's channel. The empty array [] signifies that no additional configuration is needed for the OAuth2 security mechanism.
```
